src/client/auth.js

```diff
@@ -14,9 +15,56 @@ function validateOptions (options) {
   }
 }
 
+function retry (fn, maxRetries, interval = 1000) {
+  return new Promise((resolve, reject) => {
+    let retries = 0
+    const retryFn = () => {
+      retries++
+      fn().then(resolve).catch(err => {
+        debug('retry', retries, err)
+        if (retries >= maxRetries) {
+          reject(err)
+        } else {
+          setTimeout(retryFn, interval)
+        }
+      })
+    }
+    retryFn()
+  })
+}
+
 async function realmAuthenticate (options) {
   validateOptions(options)
-  throw new Error('Not implemented')
+
+  options.authflow = new PrismarineAuth(options.username, options.profilesFolder, options, options.onMsaCode)
+
+  const api = RealmAPI.from(options.authflow, 'bedrock')
+  const realms = await api.getRealms()
+
+  debug('realms', realms)
+
+  if (!realms || !realms.length) throw Error('Couldn\'t find any Realms for the authenticating account')
+
+  let realm
+
+  if (options.realms.realmId) {
+    realm = realms.find(e => e.id === Number(options.realms.realmId))
+  } else if (options.realms.realmInvite) {
+    realm = await api.getRealmFromInvite(options.realms.realmInvite)
+  } else if (options.realms.pickRealm) {
+    if (typeof options.realms.pickRealm !== 'function') throw Error('realms.pickRealm must be a function')
+    realm = await options.realms.pickRealm(realms)
+  }
+
+  if (!realm) throw Error('Couldn\'t find a Realm to connect to. Authenticating account must be owner or have joined the Realm.')
+
+  const { address } = await retry(async () => realm.getAddress(), 5)
```

Comment:
Why the retry here?

Reply:
/join can be unstable returning 503 intermittently, retrying shortly after will return OK.

Reply:
That doesn't sound right. Is the API actually broken (not on a bad day, but all the time)? The reason prismarine-auth does retries is explicitly to enable backtracking on authentication failure, for example with server invalidated token information.

Reply:
The API is broken. From using it regularly in the past it's quite common for a join request to fail, even in the Bedrock client itself you'll see it call /join multiple times.

Reply:
OK, then it should either be documented in the prismarine-realm API that it's broken or do the handling code there. The library should handle problems like these.

Reply:
Right ok, will open a quick PR to move that there